geozero/src/csv/csv_writer.rs

```diff
@@ -0,0 +1,408 @@
+use buffering_wkt_writer::BufferingWktWriter;
+
+use std::io::Write;
+
+pub struct CsvWriter<'w, W: Write> {
+    csv: csv::Writer<&'w mut W>,
+    headers: Vec<String>,
+    has_written_first_record: bool,
+    current_row_props: Vec<String>,
+    wkt_writer: BufferingWktWriter,
+}
+
+impl<'w, W: Write> CsvWriter<'w, W> {
+    pub fn new(out: &'w mut W) -> Self {
+        let wkt_writer = BufferingWktWriter::new();
+        Self {
+            csv: csv::Writer::from_writer(out),
+            headers: vec!["geometry".to_string()],
+            has_written_first_record: false,
+            current_row_props: vec![],
+            wkt_writer,
+        }
+    }
+
+    pub fn set_dims(&mut self, dims: CoordDimensions) {
+        self.wkt_writer.dims = dims;
+    }
+}
+
+impl<W: Write> FeatureProcessor for CsvWriter<'_, W> {
+    fn dataset_begin(&mut self, _name: Option<&str>) -> Result<()> {
+        debug_assert_eq!(self.headers, &["geometry"]);
+        Ok(())
+    }
+    fn dataset_end(&mut self) -> Result<()> {
+        self.headers = vec!["geometry".to_string()];
+        Ok(())
+    }
+    fn feature_begin(&mut self, _idx: u64) -> Result<()> {
+        debug_assert!(self.current_row_props.is_empty());
+        Ok(())
+    }
+
+    fn feature_end(&mut self, _idx: u64) -> Result<()> {
+        if !self.has_written_first_record {
+            self.has_written_first_record = true;
+            self.csv.write_record(self.headers.clone())?;
+        }
+
+        let geom = self.wkt_writer.bytes();
+        self.csv.write_field(geom)?;
+        self.wkt_writer.clear();
+
+        for field in &self.current_row_props {
+            self.csv.write_field(field)?;
+        }
+        self.csv.write_record(None::<&[u8]>)?;
+        self.current_row_props.clear();
+
+        Ok(())
+    }
+    fn properties_begin(&mut self) -> Result<()> {
+        debug_assert!(self.current_row_props.is_empty());
+        Ok(())
+    }
+    fn properties_end(&mut self) -> Result<()> {
+        Ok(())
+    }
+    fn geometry_begin(&mut self) -> Result<()> {
+        debug_assert!(self.wkt_writer.bytes().is_empty());
+        Ok(())
+    }
+    fn geometry_end(&mut self) -> Result<()> {
+        Ok(())
+    }
+}
+
+impl<W: Write> PropertyProcessor for CsvWriter<'_, W> {
+    fn property(&mut self, i: usize, colname: &str, colval: &ColumnValue) -> Result<bool> {
+        // TODO: support mis-ordered properties?
+        if self.has_written_first_record {
+            assert_eq!(
+                colname,
+                &self.headers[i + 1],
+                "CSV features must all have the same column names"
+            );
+        } else {
+            self.headers.push(colname.to_string());
+        }
+
+        // TODO: support non-string colval
+        self.current_row_props.push(colval.to_string());
+        Ok(false)
+    }
+}
+
+impl<W: Write> GeomProcessor for CsvWriter<'_, W> {
+    fn dimensions(&self) -> CoordDimensions {
+        self.wkt_writer.dimensions()
```

Comment:
All this explicit delegation is kind of tedious. But I'm not sure of a better way.